### PR TITLE
Liquidity incentives

### PR DESCRIPTION
**Pull request by illia-bobyr**
TODO: Need to add a section to the `README.md`, to explain how to run the commands.

---
This change is [Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/7)

### CODE REVIEW

File: src/uniswap/incentives.ts [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+                  10 ** Number(coefficientsDigits)
+              )
+            );
+            liquidityAdjustment = event.liquidity;
+            break;
+
+          case "burn":
+            timeCoefficient = BigInt(
+              Math.floor(
+                (differenceInMilliseconds(max([to, block.timestamp]), from) /
```
Comment (gerben-stavenga) [inline src/uniswap/incentives.ts:288]: I think this should be the same as for mint.

Comment (gerben-stavenga) [reply src/uniswap/incentives.ts:288]: Overall I think we could be simpler

f(t, provider) = sum_{events: time(event) < t} (t-time(event) * deltaLiquidity(event)

then we are interested in f(t_end,provider) - f(t_start, provider)

Comment (gerben-stavenga) [reply src/uniswap/incentives.ts:288]: Oh of course forgot the binance price window restriction. So we have to iterate over the one minute time intervals but the point is there is no difference between mint/burn except for the sign of the deltaLiquidity

**Review by maxibustos — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/7)* status: 0 of 9 files reviewed, 5 unresolved discussions (waiting on @gerben-stavenga, @ilya-bobyr, and @maxibustos)

___
*[src/binance.ts, line 12 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6fNm7Kp4vlQwck-2:-Mnl6fNm7Kp4vlQwck-3:b-9r5rlb) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/binance.ts#L12)):*
> ```XML
> 
> /**
>  * Bianance API interaction
> ```

Binance
___
*[src/binance.ts, line 16 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6gnKBYNFe7xGj0zH:-Mnl6gnKBYNFe7xGj0zI:b8oq009) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/binance.ts#L16)):*
> ```XML
> 
> /**
>  * Binance support a number of price intervals.
> ```

supports
___
*[src/uniswap.ts, line 75 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6paG-WHEfg8zxOst:-Mnl6paG-WHEfg8zxOsu:b9hlvbl) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/uniswap.ts#L75)):*
> ```XML
>   if (lcNetwork === "mainnet") {
>     throw new Error(
>       "Operatoins on the mainnet is not supported at the moment." +
> ```

Operations
___
*[src/uniswap/incentives.ts, line 145 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6utD0dBdx6H1mmKW:-Mnl6utD0dBdx6H1mmKX:b-troz4g) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/uniswap/incentives.ts#L145)):*
> ```XML
>   }
> 
>   const incentivesDecimals = 16n;
> ```

Shouldn't this be 18n?

**Review by illia-bobyr — COMMENTED** [review]
> Dismissed @gerben-stavenga and @maxibustos from 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/7)* status: 0 of 9 files reviewed, all discussions resolved (waiting on @gerben-stavenga and @maxibustos)

___
*[src/binance.ts, line 12 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6fNm7Kp4vlQwck-2:-Mnlc5c5EECrCw3yYztO:b-896fix) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/binance.ts#L12)):*
<details><summary><i>Previously, maxibustos (Maxi Bustos) wrote…</i></summary><blockquote>

Binance
</blockquote></details>

Done.
___
*[src/binance.ts, line 16 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6gnKBYNFe7xGj0zH:-Mnlc6EC8iKsARteCq73:b-896fix) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/binance.ts#L16)):*
<details><summary><i>Previously, maxibustos (Maxi Bustos) wrote…</i></summary><blockquote>

supports
</blockquote></details>

Done.
___
*[src/uniswap.ts, line 75 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6paG-WHEfg8zxOst:-Mnlc1Wz7MNzwFZ5AOBi:b-896fix) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/uniswap.ts#L75)):*
<details><summary><i>Previously, maxibustos (Maxi Bustos) wrote…</i></summary><blockquote>

Operations
</blockquote></details>

Done.
___
*[src/uniswap/incentives.ts, line 145 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl6utD0dBdx6H1mmKW:-MnlcXEgDee272rk8zig:bqj1z7n) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/uniswap/incentives.ts#L145)):*
<details><summary><i>Previously, maxibustos (Maxi Bustos) wrote…</i></summary><blockquote>

Shouldn't this be 18n?
</blockquote></details>

Added a comment, explaining that it is used to decide on the precision of the conversion from BigInt to Number.
As 64 bit floats have only 15 significant digits, we should actually use a smaller number here.
___
*[src/uniswap/incentives.ts, line 288 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/7#-Mnl1ugVnsW3iM85cZxL-r1-288:-Mnle2_71SbsmdhnxFhz:bxxvoc5) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7edff57c34a644245306ba7bb43b95ed32f9732a/src/uniswap/incentives.ts#L288)):*
<details><summary><i>Previously, gerben-stavenga wrote…</i></summary><blockquote>

Oh of course forgot the binance price window restriction. So we have to iterate over the one minute time intervals but the point is there is no difference between mint/burn except for the sign of the deltaLiquidity
</blockquote></details>

Good point!
Fixed.

**Review by maxibustos — APPROVED** [review]
> Reviewed 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/7)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ilya-bobyr)

**Review by h-alnami — COMMENTED** [review]
Done

**Pull request merged**